[
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/57",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/57/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/57/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/57/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/57",
        "id": 204513173,
        "node_id": "MDU6SXNzdWUyMDQ1MTMxNzM=",
        "number": 57,
        "title": "Exception on program exit using SSL in python >= 3.4 ?",
        "user": {
            "login": "petersilva",
            "id": 394117,
            "node_id": "MDQ6VXNlcjM5NDExNw==",
            "avatar_url": "https://avatars.githubusercontent.com/u/394117?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/petersilva",
            "html_url": "https://github.com/petersilva",
            "followers_url": "https://api.github.com/users/petersilva/followers",
            "following_url": "https://api.github.com/users/petersilva/following{/other_user}",
            "gists_url": "https://api.github.com/users/petersilva/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/petersilva/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/petersilva/subscriptions",
            "organizations_url": "https://api.github.com/users/petersilva/orgs",
            "repos_url": "https://api.github.com/users/petersilva/repos",
            "events_url": "https://api.github.com/users/petersilva/events{/privacy}",
            "received_events_url": "https://api.github.com/users/petersilva/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 1,
        "created_at": "2017-02-01T07:18:01Z",
        "updated_at": "2017-02-01T21:04:59Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "I think this will hit anyone connecting to an SSL broker on python > 3.4, but not sure.   If someone else can reproduce it, that would be good.\r\n\r\nI connect to a broker, and without SSL, the code works fine.  When SSL is on, it still works fine, but as part of sys.exit() processing it prints out this:\r\n\r\nException ignored in: <bound method _AbstractTransport.\\_\\_del\\_\\_ of <amqplib.client_0_8.transport.SSLTransport object at 0x7f93e0834dd8>>\r\nTraceback (most recent call last):\r\n  File \"/home/peter/.local/lib/python3.5/site-packages/amqplib-1.0.2-py3.5.egg/amqplib/client_0_8/transport.py\", line 97, in \\_\\_del\\_\\_\r\n  File \"/home/peter/.local/lib/python3.5/site-packages/amqplib-1.0.2-py3.5.egg/amqplib/client_0_8/transport.py\", line 136, in close\r\n  File \"/home/peter/.local/lib/python3.5/site-packages/amqplib-1.0.2-py3.5.egg/amqplib/client_0_8/transport.py\", line 208, in _shutdown_transport\r\n  File \"/usr/lib/python3.5/ssl.py\", line 971, in unwrap\r\n    s = self._sslobj.unwrap()\r\n  File \"/usr/lib/python3.5/ssl.py\", line 642, in unwrap\r\n    return self._sslobj.shutdown()\r\nssl.SSLError: Underlying socket connection gone (_ssl.c:2020)\r\nb\r\n\r\nNothing in the application seems responsible for this.  It just always happens in sys.exit()\r\n\r\nI found two methods of suppressing the error:\r\n   in amqplib/client_0_8/transport.py\r\n       \r\n   I can either comment out the \\_\\_del\\_\\_() method, or\r\n   put a try/except: pass clause in close()\r\n\r\nI read a bit, and it seems that in python >= 3.4 it has gotten more consistent about calling \\_\\_del\\_\\_()\r\n( PEP442 )  but that is just a guess. \r\n\r\nI can offer a patch with either fix if one is blessed.   But I don't know whether either approach is good.  Any suggestions?\r\n\r\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/56",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/56/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/56/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/56/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/56",
        "id": 114618688,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2ODg=",
        "number": 56,
        "title": "Exception AttributeError: \"'NoneType' object has no attribute 'SHUT_RDWR'\"",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:23Z",
        "updated_at": "2015-11-02T15:35:23Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nWhich version of amqplib are you using?\namqplib==1.0.2\n\nHave you checked to see if there is a newer version in the \"Featured\nDownloads\" section of the front page of this project?\nno\n\nWhich version of Python?\n2.7.8\n\nIn my logs I see some occurences of:\nException AttributeError: \"'NoneType' object has no attribute 'SHUT_RDWR'\" in \n<bound method TCPTransport.__del__ of \n<amqplib.client_0_8.transport.TCPTransport object at 0x7feff33a2dd0>> ignored\n\nI guess that is because\n    def close(self):\n        if self.sock is not None:\n            self._shutdown_transport()\n            # Call shutdown first to make sure that pending messages\n            # reach the AMQP broker if the program exits after\n            # calling this method.\n            self.sock.shutdown(socket.SHUT_RDWR)\nin a situation where Python is shutting down and the socket module already is \nempty ... or something like that. \n\nThis is not a big deal for me - I just wanted to let you know and share the \npossible hint of what the root cause is.\n\n```\n\nOriginal issue reported on code.google.com by `kiile...@gmail.com` on 19 Apr 2015 at 4:41\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/55",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/55/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/55/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/55/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/55",
        "id": 114618686,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2ODY=",
        "number": 55,
        "title": "Roadmap?",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:22Z",
        "updated_at": "2015-11-02T15:35:22Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nAre there any plans to add support for trace frames?\n\nThanks,\nVitali\n```\n\nOriginal issue reported on code.google.com by `kir...@gmail.com` on 10 Dec 2012 at 9:32\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/54",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/54/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/54/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/54/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/54",
        "id": 114618682,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2ODI=",
        "number": 54,
        "title": "Python3 compatibility: raised exception",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 1,
        "created_at": "2015-11-02T15:35:22Z",
        "updated_at": "2015-11-02T15:35:22Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nDue to a difference in exceptions between python3 and python2 the py2to3 \ndoesn't do a perfect job.\nhttp://docs.python.org/py3k/howto/pyporting.html#capturing-the-currently-raised-\nexception\n\nFix is below. Basically stores the exception explicitly instead of leaking it.\n\n--- transport.py    2012-10-04 10:34:36.737574933 +0200\n+++ transport_fixed.py  2012-10-04 10:35:17.377255648 +0200\n@@ -74,7 +74,8 @@\n                 self.sock = socket.socket(af, socktype, proto)\n                 self.sock.settimeout(connect_timeout)\n                 self.sock.connect(sa)\n-            except socket.error, msg:\n+            except socket.error, msg_in:\n+                msg=msg_in\n                 self.sock.close()\n                 self.sock = None\n                 continue\n\n```\n\nOriginal issue reported on code.google.com by `killian....@megasoft.be` on 4 Oct 2012 at 8:39\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/53",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/53/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/53/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/53/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/53",
        "id": 114618680,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2ODA=",
        "number": 53,
        "title": "Misleading error message on socket open when hitting ulimit",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:22Z",
        "updated_at": "2015-11-02T15:35:22Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nWhich version of amqplib are you using? 1.0.2\n\nHave you checked to see if there is a newer version in the \"Featured\nDownloads\" section of the front page of this project? Yup.\n\nWhich broker are you using (RabbitMQ?) which version? RabbitMQ, 2.8.4\n\nWhich version of Python? 2.7.1\n\nWhat steps will reproduce the problem?\n1. Hit your max open file descriptors limit\n2. Try to initialize something base on _AbstractTransport\n\n\nWhat is the expected output? What do you see instead?\n\nI expect to see something like: \n\nTraceback (most recent call last):\n  File \"build/bdist.macosx-10.7-intel/egg/amqplib/client_0_8/connection.py\", line 129, in __init__\n  File \"build/bdist.macosx-10.7-intel/egg/amqplib/client_0_8/transport.py\", line 282, in create_transport\n  File \"build/bdist.macosx-10.7-intel/egg/amqplib/client_0_8/transport.py\", line 86, in __init__\nerror: [Errno 24] Too many open files\n\nBut I actually see something like:\n\n  File \"build/bdist.macosx-10.7-intel/egg/amqplib/client_0_8/connection.py\", line 129, in __init__\n  File \"build/bdist.macosx-10.7-intel/egg/amqplib/client_0_8/transport.py\", line 282, in create_transport\n  File \"build/bdist.macosx-10.7-intel/egg/amqplib/client_0_8/transport.py\", line 79, in __init__\nAttributeError: 'NoneType' object has no attribute 'close'\n\nwhich hides the real error. \n\nI've attached a patch that fixes this.\n```\n\nOriginal issue reported on code.google.com by `p...@triplefox.com` on 23 Aug 2012 at 9:23\n\nAttachments:\n- [socketclose.patch](https://storage.googleapis.com/google-code-attachments/py-amqplib/issue-53/comment-0/socketclose.patch)\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/52",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/52/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/52/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/52/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/52",
        "id": 114618677,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2Nzc=",
        "number": 52,
        "title": "Modify AbstractTransport class for supporting KeepAlive options",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:21Z",
        "updated_at": "2015-11-02T15:35:22Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nI'd like to propose a change for the AbstractTranspor class for adding a \n**kwargs parameter.\n\nThis parameter could be used to pass specific option for sockets, in particular \noptions for defining values for KEEPALIVE settings:\nsocket.TCP_KEEPIDLE\nsocket.KEEPINTVL\nsocket.KEEPCNT\n\nIf those keys are found the values are defined.\nWe could also change the code to verify if a key is passed for the SO_KEEPALIVE \noption for enable/disable it.\n\nThe aim of this change is to allow to define specific params for keep_alive \ninstead of using default OS values.\n```\n\nOriginal issue reported on code.google.com by `andrea.rosa` on 12 Jul 2012 at 3:51\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/51",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/51/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/51/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/51/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/51",
        "id": 114618675,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NzU=",
        "number": 51,
        "title": "Installation failure using Python 3",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 2,
        "created_at": "2015-11-02T15:35:21Z",
        "updated_at": "2015-11-02T15:35:21Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\n $ sudo /opt/python3.3/bin/easy_install-3.3 amqplib\nSearching for amqplib\nReading http://pypi.python.org/simple/amqplib/\nReading http://barryp.org/software/py-amqplib/\nReading http://code.google.com/p/py-amqplib/\nBest match: amqplib 1.0.2\nDownloading \nhttp://pypi.python.org/packages/source/a/amqplib/amqplib-1.0.2.tgz#md5=5c92f17fb\nedd99b2b4a836d4352d1e2f\nProcessing amqplib-1.0.2.tgz\nRunning amqplib-1.0.2/setup.py -q bdist_egg --dist-dir \n/tmp/easy_install-hkwjp6/amqplib-1.0.2/egg-dist-tmp-i5oiet\nzip_safe flag not set; analyzing archive contents...\nTraceback (most recent call last):\n  File \"/opt/python3.3/bin/easy_install-3.3\", line 9, in <module>\n    load_entry_point('distribute==0.6.24', 'console_scripts', 'easy_install-3.3')()\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 1883, in main\n    with_ei_usage(lambda:\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 1864, in with_ei_usage\n    return f()\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 1887, in <lambda>\n    distclass=DistributionWithoutHelpCommands, **kw\n  File \"/opt/python3.3/lib/python3.3/distutils/core.py\", line 148, in setup\n    dist.run_commands()\n  File \"/opt/python3.3/lib/python3.3/distutils/dist.py\", line 917, in run_commands\n    self.run_command(cmd)\n  File \"/opt/python3.3/lib/python3.3/distutils/dist.py\", line 936, in run_command\n    cmd_obj.run()\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 349, in run\n    self.easy_install(spec, not self.no_deps)\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 589, in easy_install\n    return self.install_item(spec, dist.location, tmpdir, deps)\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 619, in install_item\n    dists = self.install_eggs(spec, download, tmpdir)\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 809, in install_eggs\n    return self.build_and_install(setup_script, setup_base)\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 1086, in build_and_install\n    self.run_setup(setup_script, setup_base, args)\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/easy_install.py\", line 1075, in run_setup\n    run_setup(setup_script, args)\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/sandbox.py\", line 31, in run_setup\n    lambda: exec(compile(open(\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/sandbox.py\", line 73, in run\n    return func()\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/sandbox.py\", line 34, in <lambda>\n    {'__file__':setup_script, '__name__':'__main__'})\n  File \"setup.py\", line 47, in <module>\n    Programming Language :: Python :: Implementation :: PyPy\n  File \"/opt/python3.3/lib/python3.3/distutils/core.py\", line 148, in setup\n    dist.run_commands()\n  File \"/opt/python3.3/lib/python3.3/distutils/dist.py\", line 917, in run_commands\n    self.run_command(cmd)\n  File \"/opt/python3.3/lib/python3.3/distutils/dist.py\", line 936, in run_command\n    cmd_obj.run()\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/bdist_egg.py\", line 227, in run\n    os.path.join(archive_root,'EGG-INFO'), self.zip_safe()\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/bdist_egg.py\", line 266, in zip_safe\n    return analyze_egg(self.bdist_dir, self.stubs)\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/bdist_egg.py\", line 402, in analyze_egg\n    safe = scan_module(egg_dir, base, name, stubs) and safe\n  File \"/opt/python3.3/lib/python3.3/site-packages/distribute-0.6.24-py3.3.egg/setuptools/command/bdist_egg.py\", line 429, in scan_module\n    code = marshal.load(f);  f.close()\nValueError: bad marshal data (unknown type code)\n\n```\n\nOriginal issue reported on code.google.com by `grflana...@gmail.com` on 9 Mar 2012 at 8:31\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/50",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/50/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/50/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/50/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/50",
        "id": 114618671,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NzE=",
        "number": 50,
        "title": "Occasionally missing work",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 3,
        "created_at": "2015-11-02T15:35:20Z",
        "updated_at": "2015-11-02T15:35:21Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nAmqplib 1.0.2 with timeout patch. Python 3.2. RabbitMQ 2.7.1.\n\nI have two workers on the same server processing incoming messages and replying \nRPC-style (i.e. preserving correlation_id). Also, \nchan.basic_qos(prefetch_size=0,prefetch_count=1,a_global=False) and \nchan.basic_consume(queue=q, no_ack=False, callback=data_received)\n\nEach work item takes 200-2000ms depending on content. There is plenty of time \nbetween work items, they arrive 1-20 per second. The two workers are mainly for \nredundancy (being able to upgrade and restart without downtime).\n\nI have logging, with timestamps, in my own code both on submitting work to \nRabbitMQ and on processing it.\n\nOccasionally, I will see work packets being delayed maybe 25 seconds from being \ngiven to RabbitMQ before arriving at the client. This happens randomly and \nabout once every 5K-10K work items. After 20 seconds the client times out so \nwhile the work is eventually done, no one is interested in the result.\n\nRabbitMQ and the workers are separated by one network switch, the servers are \njust next to eachother. The Rabbit queue setup is very simple with just a \ndefault/direct exchange.\n\nI am unsure where the problem lies, but my own logging indicates it is either \nRabbitMQ or amqplib. I would be happy to log more, share code, share data or \notherwise do work to follow up on people's hunches.\n\nIf anyone is interested, the site in question is boardword.com\n\n```\n\nOriginal issue reported on code.google.com by `fgunder...@gmail.com` on 5 Mar 2012 at 10:09\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/49",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/49/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/49/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/49/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/49",
        "id": 114618667,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2Njc=",
        "number": 49,
        "title": "Tests are broken with Py3k",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162164,
                "node_id": "MDU6TGFiZWwyODExNjIxNjQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Enhancement",
                "name": "Type-Enhancement",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 4,
        "created_at": "2015-11-02T15:35:19Z",
        "updated_at": "2015-11-02T15:35:20Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nVersion: 1.0.2\n\nRunning \"python3.2 tests/client_0_8/run_all.py\" produce a syntax error because \nu\"str\" is not supported by python 3.\n\nI attached a small patch that fix this issue.\n\n\n```\n\nOriginal issue reported on code.google.com by `corentin...@gmail.com` on 4 Mar 2012 at 10:40\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/48",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/48/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/48/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/48/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/48",
        "id": 114618665,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NjU=",
        "number": 48,
        "title": "Does not recover from network downtime",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 2,
        "created_at": "2015-11-02T15:35:19Z",
        "updated_at": "2015-11-02T15:35:19Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nI am using amqplib v1.0.2 (with timeout patch) on Python 3.2.\n\nWhile the lib is in channel.wait() waiting for a message (which, of course, it \nis most of the time):\n\n- Remove power to the WLAN router of which the computer is connected.\n- Wait a few seconds\n- Plug power back in\n\nObserve: NOTHING happens. The amqplib client is now just waiting indefinitely, \nwill never exit and will never receive a message.\n\n\n```\n\nOriginal issue reported on code.google.com by `fgunder...@gmail.com` on 8 Feb 2012 at 9:52\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/47",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/47/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/47/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/47/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/47",
        "id": 114618661,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NjE=",
        "number": 47,
        "title": "Socket exception uses variable before it is declared",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 4,
        "created_at": "2015-11-02T15:35:18Z",
        "updated_at": "2015-11-02T15:35:19Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nI am using v1.0.2 (most recent as of now) on Python 3.2 with the most recent \nRabbitMQ. When there is a network issue it sometimes crashes on connect \n(amqp.Connection()) with this trace:\n\n\\connection.py\", line 129, in __init__\n    self.transport = create_transport(host, connect_timeout, ssl)\n\\transport.py\", line 297, in create_transport\n    return TCPTransport(host, connect_timeout)\n\\transport.py\", line 87, in __init__\n    raise socket.error(msg)\nUnboundLocalError: local variable 'msg' referenced before assignment\n\nI replaced the offending line\nraise socket.error(msg)\nwith:\nraise socket.error('Unable to connect')\n\n```\n\nOriginal issue reported on code.google.com by `fgunder...@gmail.com` on 7 Feb 2012 at 9:34\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/46",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/46/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/46/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/46/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/46",
        "id": 114618658,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NTg=",
        "number": 46,
        "title": "AMQPChannelException 404 no excahnge AND AMQPConnectionException 503, second 'channel.open'",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 5,
        "created_at": "2015-11-02T15:35:17Z",
        "updated_at": "2015-11-02T15:35:18Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nWhich version of amqplib are you using?\n0.6.1-1\nHave you checked to see if there is a newer version in the \"Featured\nDownloads\" section of the front page of this project?\nYes\nWhich broker are you using (RabbitMQ?) which version?\n2.6.1\nWhich version of Python?\n2.7\nWhat steps will reproduce the problem?\nIt seems some race condition, sometimes I can see the 2 exception in my log \nfile.\n\nWhat is the expected output? What do you see instead?\nThe expected behavior is to publish a message but the publishing process dies \nafter the exception\n\nPlease provide any additional information below.\nI can see those 2 exception in the log file of my application using py-amqplib \nthrough kombu library.\nThe strange thins is that the first exception is:\n<class 'amqplib.client_0_8.exceptions.AMQPChannelException'>, \nAMQPChannelException(404, u\"NOT_FOUND - no exchange \n'55f1152e2ca14745a9f85393bc24f5b4' in vhost '/nova'\", (60, 40), \n'Channel.basic_publish')\n\nand then there is the other exception:\n<class 'amqplib.client_0_8.exceptions.AMQPConnectionException'>, \nAMQPConnectionException(503, u\"COMMAND_INVALID - second 'channel.open' seen\", \n(20, 10), 'Channel.open'\n\nbut if I check in the rabbitmq I can see the exchange.\n\nUnfortunately this problem occurs intermittent and it's difficult to debug, \nmaybe the 1.0 version can have a fix to that problem?\nThanks\n\n\n```\n\nOriginal issue reported on code.google.com by `andrea.rosa` on 24 Nov 2011 at 10:09\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/45",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/45/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/45/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/45/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/45",
        "id": 114618656,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NTY=",
        "number": 45,
        "title": "connection.close() can raise \"socket.error: [Errno 57] Socket is not connected\" on Mac OS X",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:17Z",
        "updated_at": "2015-11-02T15:35:17Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nWhen an amqplib 1.0.2 connection is asked to close:\n\n1. It sends message (10, 60).\n2. It receives message (10, 61).\n3. This invokes ._close_ok().\n4. Which invokes ._do_close().\n5. Which calls .transport.close().\n6. Which attempts self.sock.shutdown(...).\n\nThis creates an exciting race condition. Which will happen first? Will the \nsocket .shutdown() method be successfully called on a still-open socket? Or \nwill the FIN packet from the server (RabbitMQ, in my case) arrive fast enough \nto mark the socket as closed before Python gets around to invoking the socket's \nshutdown() system call?\n\nOn Linux and Windows, this is an entirely safe race condition \u2014 both \noperating systems are very forgiving about calling shutdown() on a closed \nsocket, so the above sequence always works.\n\nBut on Mac OS X \u2014 a BSD variant \u2014 the above code only succeeds if the \noperating system thinks that the socket is still open when shutdown() is \ninvoked. If Python is too slow and the FIN packet arrives before that statement \ncan be reached, then OS X kills the self.sock.shutdown() statement with:\n\nsocket.error: [Errno 57] Socket is not connected\n\nAs there seems to be no way to check whether a socket is closed \u2014 nor, I \nguess, would it help, since the socket could close between such a check and the \nactual shutdown() call! So there are two possibilities here:\n\n1. Protect shutdown() with a try\u2026except that catches the socket.error, tests \nto make sure Errno is right, and ignores it if Errno matches.\n2. Ditch the shutdown() altogether. Don't all modern OS's perform a shutdown() \non a closed socket anyway? I am suspicious of the claim in the transport.py \ncomment that outgoing data could be lost; I would want to see evidence that any \nmodern TCP/IP stack behaves in that way. I cannot find a recommendation to use \nshutdown() to avoid data loss on Stack Overflow, for example.\n\nThanks so much for amqplib \u2014 I am using it with greenlet/eventlet and the \nproject is going very well!\n```\n\nOriginal issue reported on code.google.com by `brandon....@gmail.com` on 11 Nov 2011 at 9:11\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/44",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/44/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/44/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/44/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/44",
        "id": 114618655,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NTU=",
        "number": 44,
        "title": "calling basic_cancel on a closed connection is unsafe",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:17Z",
        "updated_at": "2015-11-02T15:35:17Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nWhich version of amqplib are you using?\n0.6.1\nHave you checked to see if there is a newer version in the \"Featured\nDownloads\" section of the front page of this project?\n\nYes, but code inspection leads me to believe this is still an issue\n\nWhich broker are you using (RabbitMQ?) which version?\nyes (rabbit), not sure of version\nWhich version of Python?\n2.7\nWhat steps will reproduce the problem?\n1. create a channel\n2. start consuming\n3. kill rabbit\n4. have a basic_cancel in a try:finally: around your consumer loop\n\nWhat is the expected output? What do you see instead?\n\nI'd expect the basic_cancel to just do nothing on a closed connection (or \npossibly raise an explicit exception).\n\nWhat we see is captured here:\nhttps://bugs.launchpad.net/python-oops-amqp/+bug/884539\n\"Traceback (most recent call last):\n  File \"bin/amqp2disk\", line 49, in <module>\n    oopstools.scripts.amqp2disk.main()\n  File \"/srv/lp-oops.canonical.com/cgi-bin/lpoops/src/oopstools/scripts/amqp2disk.py\", line 113, in main\n    receiver.run_forever()\n  File \"/srv/lp-oops.canonical.com/cgi-bin/lpoops/eggs/oops_amqp-0.0.3-py2.6.egg/oops_amqp/receiver.py\", line 86, in run_forever\n    self._run_forever()\n  File \"/srv/lp-oops.canonical.com/cgi-bin/lpoops/eggs/oops_amqp-0.0.3-py2.6.egg/oops_amqp/receiver.py\", line 112, in _run_forever\n    self.channel.basic_cancel(self.consume_tag)\n  File \"/srv/lp-oops.canonical.com/cgi-bin/lpoops/eggs/amqplib-0.6.1-py2.6.egg/amqplib/client_0_8/channel.py\", line 1702, in basic_cancel\n    self._send_method((60, 30), args)\n  File \"/srv/lp-oops.canonical.com/cgi-bin/lpoops/eggs/amqplib-0.6.1-py2.6.egg/amqplib/client_0_8/abstract_channel.py\", line 69, in _send_method\n    self.connection.method_writer.write_method(self.channel_id,\nAttributeError: 'NoneType' object has no attribute 'method_writer'\n\"\n\nPlease provide any additional information below.\nWe've put a if self.channel.is_open: guard around the call to basic_cancel, but \nit would be nice to have fire-and-forget here, as it seems like an obvious \nidiom to be cancelling consumption in a try:finally:\n\n```\n\nOriginal issue reported on code.google.com by `robert.c...@gmail.com` on 1 Nov 2011 at 3:54\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/43",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/43/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/43/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/43/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/43",
        "id": 114618653,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2NTM=",
        "number": 43,
        "title": "how to create a mirrored-queue ",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:17Z",
        "updated_at": "2015-11-02T15:35:17Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nhi,\nsth from http://www.rabbitmq.com/ha.html#genesis,\namqplib can not support list type? in serialization.py?\n```\n\nOriginal issue reported on code.google.com by `yuanv...@gmail.com` on 11 Oct 2011 at 4:51\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/40",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/40/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/40/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/40/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/40",
        "id": 114618635,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2MzU=",
        "number": 40,
        "title": "Connection _do_close() does not clean up if transport.close() fails",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 1,
        "created_at": "2015-11-02T15:35:14Z",
        "updated_at": "2015-11-02T15:35:15Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nUsing v1.0.0 of amqplib, Python 2.7.1, Mac OS X 10.7. \n\nIf network connectivity is lost while a Connection is open, and the Connection \nis then closed, its _do_close() method first attempts to close its transport. \nWhen that fails, an exception is raised, and the rest of the cleanup code in \n_do_close() is not run.  This leaves Channel objects with open callback \nreferences.  Since the object that created the Connection and Channel is likely \nto provide the callback for that Channel, a circular reference results and the \nobjects are never deleted.\n\nI've worked around this problem by wrapping the self.transport.close() call in \nConnection's _do_close in a try: finally block, so the cleanup code is run \nwhether the transport.close() works or not. If this is an inappropriate \nsolution, or I'm missing something, I'd like to know. Thanks!\n\n\n\n```\n\nOriginal issue reported on code.google.com by `jgarb...@gmail.com` on 24 Sep 2011 at 2:08\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/38",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/38/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/38/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/38/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/38",
        "id": 114618628,
        "node_id": "MDU6SXNzdWUxMTQ2MTg2Mjg=",
        "number": 38,
        "title": "Missing support for sequence serialization",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 3,
        "created_at": "2015-11-02T15:35:13Z",
        "updated_at": "2015-11-02T15:35:13Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nI'm enclosing a patch both in plain and in hg bundle format, which adds:\n\n1. UnitTests for sequence serialization/deserialization\n2. Sequence serialization/deserialization support\n\nWhile under 1.0 amqplib raises a sensible exception while trying to\npublish a message whose application headers contain a sequence:\n\n>>> from  amqplib import client_0_8 as amqp\n>>> cn=amqp.Connection(userid='test', password='*********', \nvirtual_host='*******')\n>>> ch=cn.channel()\n>>> msg = amqp.Message('hello world', application_headers = \n{'hdr1':['val1','val2']})\n>>> ch.basic_publish(msg,routing_key='tests', exchange='hdrXc')\nTraceback (most recent call last):\n  File \"<stdin>\", line 1, in <module>\n  File \"amqplib/client_0_8/channel.py\", line 2310, in basic_publish\n    self._send_method((60, 40), args, msg)\n  File \"amqplib/client_0_8/abstract_channel.py\", line 76, in _send_method\n    method_sig, args, content)\n  File \"amqplib/client_0_8/method_framing.py\", line 250, in write_method\n    properties = content._serialize_properties()\n  File \"amqplib/client_0_8/serialization.py\", line 531, in _serialize_properties\n    getattr(raw_bytes, 'write_' + proptype)(val)\n  File \"amqplib/client_0_8/serialization.py\", line 397, in write_table\n    raise ValueError('%s not serializable in AMQP' % repr(v))\nValueError: ['val1', 'val2'] not serializable in AMQP\n\n\nin the 0.6.1 case the ch.basic_publish seems to return, but  on subsequent \npublish() tries I get a socket error:\n\nTraceback (most recent call last):\n  File \"<stdin>\", line 1, in <module>\n  File \"/usr/lib/pymodules/python2.6/amqplib/client_0_8/channel.py\", line 2223, in basic_publish\n    self._send_method((60, 40), args, msg)\n  File \"/usr/lib/pymodules/python2.6/amqplib/client_0_8/abstract_channel.py\", line 70, in _send_method\n    method_sig, args, content)\n  File \"/usr/lib/pymodules/python2.6/amqplib/client_0_8/method_framing.py\", line 240, in write_method\n    self.dest.write_frame(2, channel, payload)\n  File \"/usr/lib/pymodules/python2.6/amqplib/client_0_8/transport.py\", line 125, in write_frame\n    frame_type, channel, size, payload, 0xce))\n  File \"<string>\", line 1, in sendall\nsocket.error: [Errno 32] Broken pipe\n\n\n```\n\nOriginal issue reported on code.google.com by `lmcatu...@gmail.com` on 12 Sep 2011 at 1:25\n\nAttachments:\n- [amqplib_sequence.diff](https://storage.googleapis.com/google-code-attachments/py-amqplib/issue-38/comment-0/amqplib_sequence.diff)\n- [amqplib_sequence.bundle](https://storage.googleapis.com/google-code-attachments/py-amqplib/issue-38/comment-0/amqplib_sequence.bundle)\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/24",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/24/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/24/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/24/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/24",
        "id": 114618599,
        "node_id": "MDU6SXNzdWUxMTQ2MTg1OTk=",
        "number": 24,
        "title": "Connection() hangs on Windows XP",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 0,
        "created_at": "2015-11-02T15:35:03Z",
        "updated_at": "2015-11-02T15:35:03Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nWhich version of amqplib are you using?\n0.6.0, 0.6.1, hg commit 174:aa7a2b480166\n\nWhich broker are you using (RabbitMQ?) which version?\nQPid 0.6 (Java version)\n\nWhich version of Python?\n2.6.4 (distributed by python.org), 2.6.5 (distributed by Cygwin)\n\nWhat steps will reproduce the problem?\n1. Start Qpid using default plain-text passwd file. \n2. Run this python code:\n    from amqplib import client_0_8 as amqp\n    amqp.Connection()\n\nWhat is the expected output? What do you see instead?\nI expect the Connection constructor to return. It never returns. Past issues \nsuggest this was fixed between 0.6.0 and 0.6.1. I confirmed that this works on \nmy linux machine. On Windows XP it still hangs.\n\n```\n\nOriginal issue reported on code.google.com by `drewpvo...@gmail.com` on 18 Nov 2010 at 11:14\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/19",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/19/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/19/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/19/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/19",
        "id": 114618591,
        "node_id": "MDU6SXNzdWUxMTQ2MTg1OTE=",
        "number": 19,
        "title": "Channel.flow  control patch by Tony Garnock-Jones works in 0.6.2devel",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 1,
        "created_at": "2015-11-02T15:34:59Z",
        "updated_at": "2015-11-02T15:34:59Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nWhich version of amqplib are you using?\n0.6.1  and 0.6.2devel with tonyg's patch for enabling channel.flow control.\n\n\nHave you checked to see if there is a newer version in the \"Featured\nDownloads\" section of the front page of this project?\n\nWhich broker are you using (RabbitMQ?) which version?\n\n1.7.2\nWhich version of Python?\n\n2.5 and 2.6\n\n\nI'm just reporting that tonyg's  http://gist.github.com/399282   patch to add \nchannel flow\ncontrol works.  I've applied it and tested it in the 0.6.2devel branch.\n\nThe only caveat is that after a basic_publish you need to issue something that \nrequires\na response from the server.  In my test issuing channel.access_request  worked \nto enable\nmy client to detect that flow control was enabled/on.\n\n\n\n```\n\nOriginal issue reported on code.google.com by `zirpubo...@gmail.com` on 24 May 2010 at 8:06\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/11",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/11/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/11/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/11/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/11",
        "id": 114618568,
        "node_id": "MDU6SXNzdWUxMTQ2MTg1Njg=",
        "number": 11,
        "title": "Support for timeout",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162155,
                "node_id": "MDU6TGFiZWwyODExNjIxNTU=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Defect",
                "name": "Type-Defect",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 12,
        "created_at": "2015-11-02T15:34:54Z",
        "updated_at": "2015-11-02T15:34:56Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nVersions: py-amqplib 0.6.1, RabbitMQ 1.5.4, Python 2.6.4\n\nI would like to perform the following:\n\n  chan.basic_consume(..)\n  heartbeat_deadline = new_deadline(...) # Compute a time instant\n  while 1:\n    next_deadline = max(0, heartbeat_deadline - time.time())\n    chan.wait(timeout = next_deadline)\n    # We get there if a message was processed,\n    # or if nothing was received for the specified amount of time.\n    # Check if we reached our deadline:\n    if time.time() > heartbeat_deadline:\n       chan.basic_publish(...)\n       heartbeat_deadline = new_deadline(...)\n\nThe purpose is to allow me to run a loop to receive new messages from\na queue synchronously, but that still allow me to send message back to\nthe server at regular interval (like heartbeat or something like that.)\n\nCould py-ampqlib support that easily? I've not checked sources in detail,\nbut I think _AbstractTransport._read(self, n) could support an additionnal\nparameter 'timeout', which would be used only for receiving the header\nof a new frame (other uses would set timeout to None) if .wait() specified\na timeout itself, using select() before the read() operation.\n\nBut maybe I misusing the library. Is that another way to obtain the same\nbehavior? I thought about writing a second process (and thus opening\nan additionnal TCP connection) but I'm not really happy with this idea.\nI would prefer to perform all the stuff from a single process.\n\n```\n\nOriginal issue reported on code.google.com by `fjolli...@gmail.com` on 21 Jan 2010 at 2:37\n",
        "performed_via_github_app": null
    },
    {
        "url": "https://api.github.com/repos/barryp/py-amqplib/issues/3",
        "repository_url": "https://api.github.com/repos/barryp/py-amqplib",
        "labels_url": "https://api.github.com/repos/barryp/py-amqplib/issues/3/labels{/name}",
        "comments_url": "https://api.github.com/repos/barryp/py-amqplib/issues/3/comments",
        "events_url": "https://api.github.com/repos/barryp/py-amqplib/issues/3/events",
        "html_url": "https://github.com/barryp/py-amqplib/issues/3",
        "id": 114618540,
        "node_id": "MDU6SXNzdWUxMTQ2MTg1NDA=",
        "number": 3,
        "title": "Heartbeats",
        "user": {
            "login": "GoogleCodeExporter",
            "id": 9614759,
            "node_id": "MDQ6VXNlcjk2MTQ3NTk=",
            "avatar_url": "https://avatars.githubusercontent.com/u/9614759?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/GoogleCodeExporter",
            "html_url": "https://github.com/GoogleCodeExporter",
            "followers_url": "https://api.github.com/users/GoogleCodeExporter/followers",
            "following_url": "https://api.github.com/users/GoogleCodeExporter/following{/other_user}",
            "gists_url": "https://api.github.com/users/GoogleCodeExporter/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/GoogleCodeExporter/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/GoogleCodeExporter/subscriptions",
            "organizations_url": "https://api.github.com/users/GoogleCodeExporter/orgs",
            "repos_url": "https://api.github.com/users/GoogleCodeExporter/repos",
            "events_url": "https://api.github.com/users/GoogleCodeExporter/events{/privacy}",
            "received_events_url": "https://api.github.com/users/GoogleCodeExporter/received_events",
            "type": "User",
            "site_admin": false
        },
        "labels": [
            {
                "id": 281162154,
                "node_id": "MDU6TGFiZWwyODExNjIxNTQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Priority-Medium",
                "name": "Priority-Medium",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162164,
                "node_id": "MDU6TGFiZWwyODExNjIxNjQ=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/Type-Enhancement",
                "name": "Type-Enhancement",
                "color": "ededed",
                "default": false,
                "description": null
            },
            {
                "id": 281162156,
                "node_id": "MDU6TGFiZWwyODExNjIxNTY=",
                "url": "https://api.github.com/repos/barryp/py-amqplib/labels/auto-migrated",
                "name": "auto-migrated",
                "color": "ededed",
                "default": false,
                "description": null
            }
        ],
        "state": "open",
        "locked": false,
        "assignee": null,
        "assignees": [],
        "milestone": null,
        "comments": 5,
        "created_at": "2015-11-02T15:34:49Z",
        "updated_at": "2015-11-02T15:34:49Z",
        "closed_at": null,
        "author_association": "NONE",
        "active_lock_reason": null,
        "body": "```\nIt looks like the client is sending a the broker a heartbeat value, but,\nbeing single threaded, I'm presuming that it's not maintaining actual\nheartbeats.\n\nWhat's the status of this feature? Should heartbeats always be 0?\n\nAs a side question, in the absence of heartbeats, what's a good/cheap way\nto actively test the broker connection to confirm we can still talk to the\nbroker? E.g. as a hack (or maybe good design, shrug :) I'm declaring a\nnon-durable queue whose sole purpose is to let me communicate with the broker.\n```\n\nOriginal issue reported on code.google.com by `g%rre.tt@gtempaccount.com` on 1 Sep 2009 at 1:25\n",
        "performed_via_github_app": null
    }
]